src/app/post-ad-component/post-ad-component.component.ts

```diff
@@ -97,17 +100,91 @@ export class PostAdComponentComponent implements OnInit {
     this.imageArray[id].imageLoaderProperty="";
   }
 
+  isValidImage(file):boolean{
+    var validFormats = ['jpg','jpeg','png'];
+    let  fName = file.name;
+    var ext = fName.substr(fName.lastIndexOf('.')+1);
+    if(ext=='') 
+      return false;
+    if(validFormats.indexOf(ext) == -1){
+        return false;
+    }
+    return true;
+  }
+
+  incrementProgressBar(id,event):void{
+    const percentDone = Math.round(100 * event.loaded / event.total);
+    this.imageArray[id].ProgressBarDispProp="";
+    console.log(`File is ${percentDone}% uploaded.`);
+    this.imageArray[id].uploadedPercent=percentDone;
+  }
+
+  getImageUrl(event):string
+  {
+    let imageUrl = "";
+    let safeUrl;    
+    console.log('File is completely uploaded!');
+    imageUrl = this.serverUrl+event.body.imageUrl;
+    console.log(imageUrl);
+    safeUrl = this.sanatizer.bypassSecurityTrustUrl(imageUrl);  // to bypass sanatization of local url
+    return safeUrl;
+  }
+
+  uploadImage(event,id)
+  {
+    this.uploadImageService.uploadImage(event.target.files[0])
+      .subscribe(
+        event => {
+            if (event.type === HttpEventType.UploadProgress) 
+            {
+              this.incrementProgressBar(id,event);
+            } 
+            else if (event instanceof HttpResponse)
+            {
+              this.imageArray[id].imageURL = this.getImageUrl(event);
+              this.productModel.imageUrls.push(event.body.imageUrl.split("/")[1]); //storing only the name of the file not the url as it may change on the server side
+              this.imageArray[id].ProgressBarDispProp="none";
+            }
+        },   
+        error=>{
+          this.removeImage(id);
+          let errMsg = "Unknown Server Error or Server Unreachable";
+          if(error.error !=null && error.error.Message != undefined)
+            errMsg = error.error.Message;
+            
+          console.log("Upload Failed\n Error: "+ errMsg);
+          alert("Upload Failed\n Error: "+ errMsg);
```

Comment:
alert?

Reply:
In case the user enters invalid image file, eg: Some  .exe renamed with .png extension would be sent to the server but the server would return a message "Invalid file format", where we should let the user know why there image is not being accepted by the server, so i directly used alerts to notify the user. I will change it to use some better way.

Reply:
display error on UI itself instead of alert.